Consolidate overall Global Styles mechanics in the server

oandregal opened the pull request:
This PR seeks to consolidate the overall server-side mechanism for Global Styles. Continues work from https://github.com/WordPress/gutenberg/pull/19883

1) It adds the user's Global Styles to the resolver, so it now works using core, theme, and user data. It stores the user's data using a Custom Post Type, so we can leverage existing editor mechanisms (undo/redo, draft/publish status for Global Styles, etc). This also presumes we want Global Styles to be tied to the theme so when theme changes we want to start with a new blank slate for Global Styles, which is implemented by using the theme's textdomain as part of the CPT's post name.

2) It exposes to the editor a couple of variables necessary for live-editing the styles: the ID of the entity that holds the user's styles + the base styles (core+theme). How this works will be demonstrated in subsequent PR. Note this is made to work only with the Beta Site Editor.

3) As per [past conversations](https://github.com/WordPress/gutenberg/pull/19883), it flattens the Global Styles tree stripping out the block variables. It looks like this:

```js
{
	'color': {
		'primary': '<primary color>',
		'background': '<background color>',
		'text': '<text color>',
	},
        'line-height': '<line height>'
}
```

which will be converted to this CSS rule (see [discussion about syntax](https://github.com/WordPress/gutenberg/pull/20047#discussion_r377547790)):

```js
:root {
	--wp--color--primary: '<primary color>';
	--wp--color--background: '<background color>';
	--wp--color--text: '<text color>';
        --wp--line-height: '<line height>';
}
```

Note that the names and variables we expose are still in flux, so take the current ones as only valid for demo purposes.

## How to test

- Install and enable one of the block-based themes (find one at [theme-experiments](https://github.com/WordPress/theme-experiments)). Instructions here assume Twenty Twenty Blocks.
- Check out this PR and `npm install && npm run build`.
- Create a **void** file named `experimental-theme.json` in your current theme's folder. For example, `wp-content/themes/twentytwenty-blocks/experimental-theme.json`.
- Got to Gutenberg > Experiments and enable "Full Site Editing".
- Go to Gutenberg > Site Editor.

**1) Check that core's Global Styles work:**

- Add a button and verify that its background color is `#52accc`.

**2) Check that theme's Global Styles override core's if present:**

- Add the following into the `experimental-theme.json` file created before.

```json
{
	"color": {
		"primary": "hotpink"
	}
}
```

- Reload the Site Editor.
- Add a button and verify that the button's color is now `hotpink`.

**3) Check that user's Global Styles override theme's if present:**

- Insert in the database a new post with the following data [1]:
  - post_type='wp_global_styles'
  - post_name='wp-global-styles-THEME-TEXTDOMAIN' (ex: 'wp-global-styles-twentytwenty-blocks')
  - post_content='{"color":{"primary":"blue"}}'
  - post_status='publish'
- Reload the Site Editor.
- Add a button and verify that the button's color is now `blue`.

---

[1] Make sure there isn't another post with the same post_name; if there is one, delete it first. Find below some ways you can add the required data for testing purposes. This is for the experimental Twenty Twenty Blocks theme, change to use your active theme's text-domain as required.

**wp-cli** -- you may need to provide other connection parameters depending on your setup:

```sh
wp post delete $(wp post list --post_type='wp_global_styles' --format=ids) --force
wp post create
    --post_type='wp_global_styles'
    --post_name='wp-global-styles-twentytwenty-blocks'
    --post_content='{"color":{"primary":"blue"}}'
    --post_status='publish'
```

**PHP** -- paste the following code within the `gutenberg_experimental_global_styles_enqueue_assets` when it comes to test user's Global Styles:

```php
	// Delete existing CPT, if any.
	$gs_cpt_list = wp_get_recent_posts( [
		'post_type' => 'wp_global_styles',
	] );
	foreach ( $gs_cpt_list as $gs_cpt ) {
		wp_delete_post( $gs_cpt['ID'] );
	}

	// Insert new one with required data.
	wp_insert_post( [
	    'ID'           => 0,
	    'post_content' => json_encode( [
	        'color'=> [
	            'primary' => 'blue',
	        ],
	    ] ),
	    'post_type'   => 'wp_global_styles',
	    'post_name'   => 'wp-global-styles-twentytwenty-blocks',
	    'post_status' => 'publish',
	] );
```

oandregal committed:
Warm up changes

oandregal committed:
Use same pattern name

oandregal committed:
Extract get_from_file function

oandregal committed:
Add placeholder to retrieve user global styles

oandregal committed:
Simplify css vars retrieval

oandregal committed:
Better docs

oandregal committed:
Better docs

oandregal committed:
Extract resolver

oandregal committed:
Simplify global styles tree processing

oandregal committed:
Fix concatenation

oandregal committed:
Simplify to global css custom prop

oandregal committed:
Register CPT to store user global styles

oandregal committed:
Add data for client within block-editor settings

oandregal committed:
Take into account theme support

oandregal committed:
Pass core&theme global styles to client

oandregal committed:
Use gutenberg_experimental_global_styles_ as function prefix

oandregal committed:
Retrieve user data from CPT

oandregal committed:
Make user data to be an array

oandregal committed:
Filter by post date and slug/name

oandregal committed:
Cool down refactorings

oandregal committed:
Create wp_global_styles draft in editor if there is no CPT

oandregal committed:
Centralize theme support check

oandregal committed:
Better docs

oandregal committed:
Consolidate block settings names

oandregal committed:
We should check theme support at later stages

oandregal committed:
Fix typo

oandregal committed:
Follow post name convention: use dash instead of underscore

nosolosw commented:
I just realized we didn't hide the GlobalStyles code behind a experimental flag. I'll add that in this PR.

oandregal committed:
Make linter happy

oandregal committed:
Only load Global Styles if experiment is enabled

oandregal committed:
Add Global Styles setting to experiments page

ItsJonQ commented:
Heya @nosolosw !! Thank you for working on this 😍 

I walked through the steps you outlined:

* [x] Check out this PR and npm install && npm run build.
* [x] Create a void file named experimental-theme.json in your current theme's folder.
* [x] Enable the "Global Styles" in the experiments page.
* [x] Create a new post with a block button, publish, and go to the front end.

However, I wasn't able to see the Global styles stuff render. No `wp-gs` className being added, no `hotpink` button.

Is there something I need to do in addition to this?

Note: Testing locally on fresh Docker install and TwentyTwenty theme.

Thanks!

nosolosw commented:
Nope, that's all you need! Just tested with a brand new site, fresh clone PR, and the steps above: it worked fine for me.

If there is no `wp-gs` or `#global-styles-inline-css` stylesheet it means GlobalStyles isn't active: either the `experimental-theme.json` couldn't be found in the top-level dir of the theme or the experiment flag isn't enabled.

As per the CLI error. It looks like you need to provide some extra info (user/pass)? If you use the env bundled with gutenberg it may be more convenient to do `npm run env cli` instead. An alternative to CLI if that gets cumbersome would be pasting the following code within the `gutenberg_experimental_global_styles_enqueue_assets` function for testing user's Global Styles:

```php
	// Delete existing CPT, if any.
	$gs_cpt_list = wp_get_recent_posts( [
		'post_type' => 'wp_global_styles',
	] );
	foreach ( $gs_cpt_list as $gs_cpt ) {
		wp_delete_post( $gs_cpt['ID'] );
	}

	// Insert new one with required data.
	wp_insert_post( [
	    'ID'           => 0,
	    'post_content' => json_encode( [
	        'color'=> [
	            'primary' => 'blue',
	        ],
	    ] ),
	    'post_type'   => 'wp_global_styles',
	    'post_name'   => 'wp-global-styles-twenty-twenty', // Note this is for TwentyTwenty!
	    'post_status' => 'publish',
	] );
```

oandregal committed:
Use FSE experiment flag instead of dedicated Global Styles one

nosolosw commented:
I saw that we were using the Full Site Editing flag in https://github.com/WordPress/gutenberg/pull/20061 so I thought it'd be best to use it here as well. Changed in https://github.com/WordPress/gutenberg/pull/20047/commits/f9d90e036a1fe4ee65371c92f70258bce9343a71 and updated the instructions accordingly.

ItsJonQ commented:
@nosolosw Hallooo!! Thank you for your support 🙏 
Unfortunately, I'm still not able to get it working on my end

I've recorded a video here walking through the instructions:

https://www.loom.com/share/0635e0c69b1042129e09a12f555920e3

Hopefully this helps with some debugging?

ItsJonQ commented:
@nosolosw Hallo!!!

Just tested this again, and it's working as expected! I have no idea why it wasn't working for me previously. Thank you so much for your efforts in helping me debug! 🤗 

It was most likely my local Docker environment. My machine has been having issues lately.
I reinstalled all Docker things 2 days ago, and it's been pretty stable since.

I left a comment for something to fix. After that, I would say this is 👍 from me!

nosolosw commented:
> I left a comment for something to fix.

@ItsJonQ I don't seem to find any comment in the PR?

ItsJonQ reviewed (changes requested):
@nosolosw Oops! I forgot to hit Request changes! 🙇

nosolosw commented on lib/global-styles.php:
```diff
@@ -1,30 +1,37 @@
 			$result = array_merge(
 				$result,
-				gutenberg_get_css_vars( $value, $new_key, false )
+				gutenberg_experimental_global_styles_get_css_vars( $value, $new_key . '-' )
```
Ah! But this was when we had different levels (globals vs blocks) so it was useful to differentiate the element part (global/block name) vs the CSS custom property name. Now that we don't have the element part (everything is prefixed by `--wp-`), is the `--` still useful?

nosolosw replied:
I mean, now we have `--wp-color-primary`. Would we want to have `--wp--color-primary` instead?

nosolosw replied:
Updated at https://github.com/WordPress/gutenberg/pull/20047/commits/7efa53e9ff73a472f4887aeb7ef763578eb72626 and https://github.com/WordPress/gutenberg/pull/20047/commits/798a0d3e4fd768689fd89175da8b5a824fd52c23

---

@jorgefilipecosta and I talked this over and I want to write down the trade-offs of the current approach so we have them documented somewhere until we find a better place.

Jorge's rationale for suggesting the double dash is this: in scoped styles (we store the CSS custom properties as block attributes) we need to deserialize the string into an object for editor things. In that case, we won't be able to distinguish between a valid CSS property name (ex: `line-height`) and a token we add (`background-color` as the resulting name of flattening the object `{background: { color: 'red' }}`). Let's work through an example. We've got this string we want to deserialize back into an object:

```
--wp-background-color: 'red';
--wp-line-height: 1.2;
```

How do we know which object to deserialize into without information about the object keys/CSS property names?

This is what we want:
```
{
    background: {
        color: 'red',
    },
    line-height: 1.2
}
```

A simple parser that creates one level per dash would return (which is not what we want, so we have to make the parser a bit more complex): 

```
{
    background: {
        color: 'red',
    },
    line {
        height: 1.2,
    }
}
```

By using double dashes the deserialization parser is simpler (each object level is separated by `--`):

```
--wp--background--color: 'red';
--wp--line-height: 1.2;
```

Personally, I'm not a fan of this implementation detail leaking into the API design/naming schema. However, I also think this is a fair approach to simplify the parser and get us going. The underlying assumption is that we store this information only once in the block attribute as a string, not as an object: when we cross that river, we'd have the opportunity to reevaluate this and change if necessary.

oandregal committed:
Update token

oandregal committed:
Adapt CSS custom properties to new convention

oandregal committed:
Use theme textdomain

this gives a well-known string and avoids the need
to parse the theme name for spaces, etc.

nosolosw commented:
This PR is ready to land, although it only works in the block-editor. We're going to integrate Global Styles in the `edit-site` editor, for which we need to add support for some things in `edit-site`:

- hook admin_body_class: we use this to add the `wp-gs` class. Although the hook is fired, we need a way to identify the `edit-site` context server-side (something akin to `$current_screen->is_block_editor`).
- hook block settings: doesn't fire for `edit-site` context.
- hook enqueue_block_assets: doesn't fire for `edit-site` context.

I'm going to work on the `edit-site` things separately, so we can land this PR without being blocked.

oandregal committed:
Detect whether we are in the site editor context.

oandregal committed:
Check for array key presence

oandregal committed:
Fire block settings hook in edit-site

oandregal committed:
Enqueue assets in a way that works for edit-site

oandregal committed:
Add some editor checks: only load if edit-site

nosolosw commented:
I was able to make this work for `edit-site` in a decoupled and simple way, so I went ahead and added the code in this branch. This works now only with `edit-site` and frontend.

ItsJonQ reviewed (approved):
🚀 @nosolosw Awesome work!!!  Thank you for creating and iterating on this! Not only that, but you've also added it to FSE "Edit Site".

Testing locally, and it's working as outlined in your instructions.

From a code perspective, the mechanics of the PHP updates make sense of me.

ItsJonQ commented:
@jorgefilipecosta If you're around, it would be great to get your thoughts on the PHP implementation 🙏 . Thank you!!

oandregal committed:
Remove unrelated changes

This has been probably inadvertenly added in a rebase.

nosolosw commented on lib/experiments-page.php:
```diff
@@ -84,6 +84,17 @@ function gutenberg_initialize_experiments_settings() {
 			'id'    => 'gutenberg-full-site-editing-demo',
 		)
 	);
+	add_settings_field(
```
This is unrelated to this PR, probably inadvertently added in a rebase. Will remove.

nosolosw replied:
Removed in https://github.com/WordPress/gutenberg/pull/20047/commits/6f1e87b1e041921dbc9e5a0582dd52920db71fbd

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
+function gutenberg_experimental_global_styles_get_user_cpt( $post_status_filter = array( 'publish' ), $should_create_draft = false ) {
+	$user_cpt         = array();
+	$post_type_filter = 'wp_global_styles';
+	$post_name_filter = 'wp-global-styles-' . strtolower( wp_get_theme()->get( 'TextDomain' ) );
```
I think as an id of the theme we should use get_stylesheet() and not the 'TextDomain'. I think the database setting the WordPress core stores in the database to identify the current theme is the stylesheet.

nosolosw replied:
The [template parts](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/template-part/index.php#L22) use the `TextDomain` for this, so I thought we may want to use the same mechanism for API consistency. I'm happy to change it to something else, but I guess it makes sense to update template parts as well?

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
+		$cpt_post_id = wp_insert_post(
+			array(
+				'post_content' => '{}',
+				'post_status'  => 'draft',
```
I think we may have a bug, in gutenberg_experimental_global_styles_get_user we are querying published posts, but we insert a draft post. On the next execution, a published post will still not exist and we will probably create another one.
I think we may be able to create the post with the published status ( the empty object will not do anything);

nosolosw replied:
Note the optional parameter `$should_create_draft`, which is false by default. We only insert a draft if there is none in the editor call `gutenberg_experimental_global_styles_get_user_cpt_id`. That's the only place we need a pre-existing CPT: the entities API only works with pre-existing objects, can't create new ones from the client-side.

jorgefilipecosta commented on lib/global-styles.php:
```diff
@@ -34,98 +44,299 @@ function gutenberg_get_css_vars( $global_styles_branch, $prefix = '', $is_start
-add_action( 'enqueue_block_assets', 'gutenberg_enqueue_global_styles_assets' );
 
 /**
- * Adds class wp-gs to the frontend body class if the theme defines a experimental-theme.json.
```
If we unconditionally add the wp-gs class we will break the styles of some blocks.
```
.wp-gs .wp-block-button__link:not(.has-background) {
	background-color: var(--wp--color--primary);
}
```

As the class is added even if the theme does not support global styles a theme that was changing the button color using the following code will stop working:
```
 .wp-block-button__link {
	background-color: red;
}
```

I think the whole idea of adding wp-gs was to mark themes that support global styles if we unconditionally add the class we would be able to not add it at all as well.
I know that for now we only add if the experiment is enabled but one day we remove the experiment and in that case, we would unconditionally add the class.
I think we should add the class when the experiment is enabled and when the theme supports global styles.

nosolosw replied:
We already do that! The class is only added when the experiment is enabled, the theme has support, and the editor is site editor (for back-end, doesn't apply to front-end).

jorgefilipecosta commented on packages/block-library/src/button/style.scss:
```diff
@@ -37,7 +37,7 @@ $blocks-button__height: 56px;
 
 .wp-gs .wp-block-button__link:not(.has-background) {
-	background-color: var(--wp-block-core-button--color--background, var(--wp-color--primary, $dark-gray-700));
+	background-color: var(--wp--color--primary);
```
Should we pass $dark-gray-700 as a fallback?

nosolosw replied:
Note that `--wp--color--primary` is going to always have a value (user-defined, the theme default, or core default), so the fallback won't be necessary.

jorgefilipecosta reviewed (commented):
Hi @nosolosw nice work here 👍  generally the changes seem good and I think they are going in the right direction.
